# STAT 207 Lab Assignment 2 - Individual Part - [25 total points]

## Dataframe Manipulation and  Cleaning

Due: Tuesday, September 6, 11:59pm CST

<hr>


## <u>Case Study</u>: Steam Game Play and Purchase Analysis
Steam is the world's most popular PC Gaming hub. With a massive collection that includes everything from AAA blockbusters to small indie titles, great discovery tools can be super valuable for Steam. We would like to use the following dataset below to learn more about Steam player behavior.

The dataset we will be analyzing is comprised of a random sample of 500 steam users and their game play and purchase behaviors. It has the following columns:
* **user_id**
* **game_name**,
* **activity**:
    - purchase: indicating that the user has *purchased* the corresponding game
    - play: indicating that the user has *played* the corresponding game (for at least some amount of time.) 
* **hours_played_if_play**:
    - if the row corresponds to a 'play' activity, this number represents the number of hours the user has played the game
    - if the row corresponds to a 'purchase' activity, this number is always a 1 (and means nothing... it's a placeholder).

<hr>


### <u>Research Goals</u>

In this analysis, we will consider a game's **"all-star" player** to be the person who had played this game for the most amount of hours.

1. Identify any missing values in the dataset and drop any rows that have a missing value.
2. Learn more about each game's "all-star" player.
3. Learn more about the "all-star" players in this dataset.


### <u>Points</u>
<p>&nbsp;</p>
<table style="border: none;border-collapse: collapse;width:123pt;">
    <tbody>
        <tr>
            <td style="color:black;font-size:15px;font-weight:700;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:general;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;width:72pt;">Problem</td>
            <td style="color:black;font-size:15px;font-weight:700;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:general;vertical-align:bottom;border:.5pt solid windowtext;border-left:none;width:51pt;">Points</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">3</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">4</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">5</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">6</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">7.1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">7.2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">8.1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">8.2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">8.3</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">9.1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">9.2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">2</td>
        </tr>
    </tbody>
</table>

<hr>

### <u>Tutorial you Will Need</u>: Utilizing the *.shape* attribute

The **.shape** attribute for a given dataframe returns a **tuple** of two values, the first value in the tuple is the number of rows, the second value is the number of columns.

In [47]:
import pandas as pd
demo_df=pd.DataFrame({'a': [1,2,3], 'b': [4,5,6]})
demo_df

a  b
0  1  4
1  2  5
2  3  6

We can index a tuple with 0 and 1 indices to extract the number of rows and columns from the dataframe separately.

In [48]:
#Number of rows and columns
demo_df.shape

(3, 2)

In [3]:
#Number of rows
demo_df.shape[0]

3

In [4]:
#Number of columns
demo_df.shape[1]

2

<hr>

### Imports

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set

<function seaborn.rcmod.set(*args, **kwargs)>

### 1. Read the csv
First, read the **steam_sample.csv** file into a dataframe and display the first five rows of the dataframe.

In [6]:
df = pd.read_csv('steam_sample.csv')
df.head()

user_id  game_name  activity hours_played_if_play
0  308653033   Unturned  purchase                    1
1  308653033   Unturned      play                  0.6
2  308653033  theHunter  purchase                    1
3  144004384     Dota 2  purchase                    1
4  144004384     Dota 2      play                   22

### 2. How many rows does the dataframe have? (Use Python code to show this).

In [5]:
df.shape

(7806, 4)

7,806 rows

### 3. Data cleaning and data type checking

We plan to analyze the **"hours_played_if_play"** variable in this assignment. In order to ensure that our subsequent functions work appropriately on this numerical variable, you should do the following.
1. Identify any values (if any) that have been encoded in the csv to represent a **missing value**. 
2. Make sure that the values in your **"hours_played_if_play"** variable are represented as numerical values (either float64 or int64 object types). If not, fix this using the methods we discussed in class.
3. Delete any rows in your dataframe that have missing values. Remember, some "missing values", in the way that they're currently encoded in the csv may not be initially detectable by Python. It is your job to figure out what these missing values are and make it so Python does detect them. And then delete any rows that have them.

In [7]:
df.isna().sum()

user_id                 0
game_name               0
activity                0
hours_played_if_play    0
dtype: int64

In [8]:
df.dtypes

user_id                  int64
game_name               object
activity                object
hours_played_if_play    object
dtype: object

In [9]:
df['hours_played_if_play'].unique()

array(['1', '0.6', '22', '1028', '1008', '148', '108', '72', '36', '35',
       '32', '21', '16', '15.8', '8.6', '7.8', '7.3', '3.1', '1.9', '1.7',
       '1.1', '0.4', '153', '63', '26', '1.4', '639', '479', '70', '65',
       '33', '30', '19.8', '16.2', '11.3', '4.2', '3.9', '2.3', '0.8',
       '0.7', '0.5', '0.3', '396', '227', '13.4', '12.6', '11.2', '10.1',
       '2.4', '210', '1.2', '0.2', '13.2', '48', '110', '0.1', '67',
       '429', '5.5', '61', '1.6', '18.3', '9.9', '4.7', '1714', '441',
       '197', '147', '117', '86', '73', '49', '46', '31', '24', '20',
       '19.7', '18.2', '14.2', '11.6', '9.8', '9.7', '8.4', '6.5', '4.8',
       '4.1', '3.7', '2.9', '2.7', '2.1', '222', 'unknown', '14.9',
       '14.1', '83', '11.1', '3.2', '395', '251', '9.3', '7.4', '54',
       '34', '1.8', '99', '98', '96', '29', '27', '23', '19.1', '18.7',
       '17.5', '17', '16.6', '14.7', '13.3', '10.7', '10.2', '10', '9.6',
       '9.5', '8', '7.9', '7.6', '7', '6.6', '6', '5.8', '5.6', '5

In [11]:
missing_values = ["unknown"]
df = pd.read_csv('steam_sample.csv', na_values=missing_values)
df['hours_played_if_play'].unique()

array([1.0000e+00, 6.0000e-01, 2.2000e+01, 1.0280e+03, 1.0080e+03,
       1.4800e+02, 1.0800e+02, 7.2000e+01, 3.6000e+01, 3.5000e+01,
       3.2000e+01, 2.1000e+01, 1.6000e+01, 1.5800e+01, 8.6000e+00,
       7.8000e+00, 7.3000e+00, 3.1000e+00, 1.9000e+00, 1.7000e+00,
       1.1000e+00, 4.0000e-01, 1.5300e+02, 6.3000e+01, 2.6000e+01,
       1.4000e+00, 6.3900e+02, 4.7900e+02, 7.0000e+01, 6.5000e+01,
       3.3000e+01, 3.0000e+01, 1.9800e+01, 1.6200e+01, 1.1300e+01,
       4.2000e+00, 3.9000e+00, 2.3000e+00, 8.0000e-01, 7.0000e-01,
       5.0000e-01, 3.0000e-01, 3.9600e+02, 2.2700e+02, 1.3400e+01,
       1.2600e+01, 1.1200e+01, 1.0100e+01, 2.4000e+00, 2.1000e+02,
       1.2000e+00, 2.0000e-01, 1.3200e+01, 4.8000e+01, 1.1000e+02,
       1.0000e-01, 6.7000e+01, 4.2900e+02, 5.5000e+00, 6.1000e+01,
       1.6000e+00, 1.8300e+01, 9.9000e+00, 4.7000e+00, 1.7140e+03,
       4.4100e+02, 1.9700e+02, 1.4700e+02, 1.1700e+02, 8.6000e+01,
       7.3000e+01, 4.9000e+01, 4.6000e+01, 3.1000e+01, 2.4000e

In [12]:
df.dtypes

user_id                   int64
game_name                object
activity                 object
hours_played_if_play    float64
dtype: object

In [13]:
df.isna().sum()

user_id                 0
game_name               0
activity                0
hours_played_if_play    8
dtype: int64

In [14]:
df = df.dropna()

In [15]:
df.shape

(7798, 4)

In [17]:
df.head()

user_id  game_name  activity  hours_played_if_play
0  308653033   Unturned  purchase                   1.0
1  308653033   Unturned      play                   0.6
2  308653033  theHunter  purchase                   1.0
3  144004384     Dota 2  purchase                   1.0
4  144004384     Dota 2      play                  22.0

### 4. Number of Dropped Rows 

How many rows with missing values did you delete from your original dataframe?

In [16]:
df.shape

(7798, 4)

8 rows were deleted

### 5. Separate the dataframes. 

Next, in this analysis we would like to analyze the game purchase information and game play information separately. First create two new dataframes:
* one that is comprised of the purchase rows
* the other that is comprised of the play rows.

In [122]:
df_purchase = df[df['activity'] == "purchase"]
df_purchase = df_purchase.reset_index(drop=True)
df_purchase.head()

user_id                        game_name  activity  hours_played_if_play
0  308653033                         Unturned  purchase                   1.0
1  308653033                        theHunter  purchase                   1.0
2  144004384                           Dota 2  purchase                   1.0
3   54103616  Counter-Strike Global Offensive  purchase                   1.0
4   54103616                   Counter-Strike  purchase                   1.0

In [119]:
df_play = df[df['activity'] == "play"]
df_play = df_play.reset_index(drop=True)
df_play.head()

user_id                        game_name activity  hours_played_if_play
0  308653033                         Unturned     play                   0.6
1  144004384                           Dota 2     play                  22.0
2   54103616  Counter-Strike Global Offensive     play                1028.0
3   54103616                   Counter-Strike     play                1008.0
4   54103616                      Left 4 Dead     play                 148.0

### 6. Percentage of Game Purchases that are Actually Played

What percentage of all game purchases in this dataset are actually played by the player that purchased them?

*You can assume that a game and user id that has a 'play' row ALSO MUST have a 'purchase' row (but not the other way around).*

**For full credit, use automated Python code to find your answer. (Ie. don't put something like '30/50', where you found the 30 and 50 from the output of other code.)**

In [120]:
len(df_play)/len(df_purchase)

0.5827075299370814

### 7. Top Game Purchaser

Next, we'd like to learn more about the people who *purchase* the most amount of games.

#### 7.1. Most Amount of Games Purchased

What is the user id of the player who has purchased the most amount of games in this dataset, and how many games have they purchased?

In [121]:
dfp1 = df_purchase.groupby('user_id').agg(sum).reset_index()
dfp1

user_id  index  hours_played_if_play
0       683019  38516                  13.0
1      1665511  49532                   8.0
2      4877411  24087                  12.0
3      4897767  49467                   8.0
4      5600822  74429                  12.0
..         ...    ...                   ...
494  306636026   5380                   1.0
495  306950369   5679                   1.0
496  308468736  16245                   3.0
497  308653033      2                   2.0
498  309265377  15454                   2.0

[499 rows x 3 columns]

In [123]:
dfp1.sort_values("hours_played_if_play",ascending=False).reset_index(drop=True)

user_id    index  hours_played_if_play
0     48798067  1854885                 399.0
1     99077905   164673                 177.0
2     43684632   301839                 172.0
3     88035166   431970                 162.0
4    118852041   602063                 152.0
..         ...      ...                   ...
494  120055850     1067                   1.0
495  116000122     5610                   1.0
496  200714421     1937                   1.0
497  201269065     3262                   1.0
498  159034250     6771                   1.0

[499 rows x 3 columns]

48798067

#### 7.2. How many games has this user actually played?

What percentage of the games did this player from 7.1 actually play?

In [124]:
len(df_play[df_play["user_id"] == 48798067]) / len(df_purchase[df_purchase["user_id"] == 48798067])

0.6365914786967418

## 8. Each Game's "All-Star" Player

Next, we'd like to identify each game's "all-star" player, which in this analysis we will consider to be the player that has played the game the longest.

### 8.1. Each Game's Highest Play Time

In a single dataframe below, display the highest amount of time that each game was played.

For full credit, your final resulting dataframe should:
1. Be comprised of *two columns*
 * game_name
 * hours_played_if_play
 
2. And should have an index going from 0,1,2,... etc.

In [153]:
dfhpt = df_play.groupby("game_name").agg("sum").reset_index()
dfhpt = dfhpt.iloc[:,[0,2]]
dfhpt

game_name  hours_played_if_play
0    1... 2... 3... KICK IT! (Drop That Beat Like a...                   0.2
1                18 Wheels of Steel American Long Haul                   5.8
2                                               3DMark                   1.3
3                                          60 Seconds!                   1.6
4                                        7 Days to Die                 249.0
..                                                 ...                   ...
915                            Zombies Monsters Robots                  19.0
916                                              hocus                   0.1
917                                                 oO                   0.2
918                                       sZone-Online                   4.3
919                                          theHunter                  13.2

[920 rows x 2 columns]

### 8.2. Ranking the Games by Highest Maximum Game Play Time

Sort your final resulting dataframe from 8.1 by the number of hours that the game was played by the all-star player *in descending order*.

Overwrite your dataframe, so it remains sorted in this order.

In [161]:
dfhpt = dfhpt.sort_values("hours_played_if_play",ascending=False).reset_index(drop=True)
dfhpt

game_name  hours_played_if_play
0                                  Dota 2               31965.8
1         Counter-Strike Global Offensive               18345.2
2              Sid Meier's Civilization V               14490.0
3                          Counter-Strike               10273.1
4              The Elder Scrolls V Skyrim                3737.2
..                                    ...                   ...
915                              Shan Gui                   0.1
916                           Remember Me                   0.1
917  Star Wars Jedi Knight Dark Forces II                   0.1
918                   The Showdown Effect                   0.1
919          The Way of Life Free Edition                   0.1

[920 rows x 2 columns]

### 8.3. Find the User Ids of each Games "All-Star" Player

Finally, display a dataframe that shows the **user id** and the **hours played** of each corresponding **game's** all-star player(s).

**For full credit, your final result should not contain an 'activity' column. It should just have a user id, hours played, and game column.**

In [182]:
dfasp = df_play.groupby("game_name").agg(max).reset_index()
dfasp = dfasp.iloc[:,[0,1,3]]
dfasp

game_name    user_id  \
0    1... 2... 3... KICK IT! (Drop That Beat Like a...   78560022   
1                18 Wheels of Steel American Long Haul   43684632   
2                                               3DMark   48798067   
3                                          60 Seconds!   95780749   
4                                        7 Days to Die  174287424   
..                                                 ...        ...   
915                            Zombies Monsters Robots  118852041   
916                                              hocus  195072781   
917                                                 oO  118852041   
918                                       sZone-Online  294834367   
919                                          theHunter  201708562   

     hours_played_if_play  
0                     0.2  
1                     5.8  
2                     1.3  
3                     1.6  
4                   175.0  
..                    ...  
915                  19.0  
916                   0.1  
917                   0.2  
918                   2.3  
919                   8.6  

[920 rows x 3 columns]

## 9. Learning about the All-Star Players

Is it possible that some players are an "all-star" player for *multiple games* in this dataset? Let's learn more.


### 9.1. All-Star Player Game Count

Considering the all-star players in the dataframe from 8.3, which player is the all-star for the *most amount of games*? And how many games are they all-star player for?

In [191]:
dfaspm = dfasp.sort_values("user_id",ascending=False).reset_index(drop=True)
dfaspm["user_id"].mode()
dfaspm

game_name    user_id  hours_played_if_play
0                          Brawlhalla  309265377                   7.4
1                    MicroVolts Surge  309265377                   2.4
2                            Unturned  308653033                 138.0
3                         Magic Duels  308468736                  34.0
4                         War Thunder  308468736                 458.0
..                                ...        ...                   ...
915  DUNGEONS - Steam Special Edition   16084897                   1.0
916                   Shattered Union   16084897                 116.0
917                   Dementium II HD   16084897                   0.7
918                    Revolution Ace   14962163                   0.2
919            Speedball 2 Tournament   10869571                   0.4

[920 rows x 3 columns]

In [193]:
check = dfaspm[dfaspm["user_id"] == 48798067]
check

game_name   user_id  \
731                                            3DMark  48798067   
732  The Chronicles of Riddick Assault on Dark Athena  48798067   
733                               Zombie Panic Source  48798067   
734                                        Hero Siege  48798067   
735                    The Elder Scrolls IV Oblivion   48798067   
..                                                ...       ...   
838                                     Another World  48798067   
839               Joe Dever's Lone Wolf HD Remastered  48798067   
840                             Kane & Lynch Dead Men  48798067   
841                           Kane & Lynch 2 Dog Days  48798067   
842              Lara Croft and the Guardian of Light  48798067   

     hours_played_if_play  
731                   1.3  
732                  15.6  
733                   4.2  
734                   8.1  
735                  59.0  
..                    ...  
838                   2.2  
839                  14.3  
840                   3.2  
841                   4.3  
842                   0.3  

[112 rows x 3 columns]

### 9.2. Top All-Star Player Stats

For the top all-star player that you identified in 9.1 above, what percentage of the *games that they have actually played* are they an all-star player for?

**Use automated Python code to help you answer this question. In other words, your final answer should not look something like '201/300', where you got 201 and 300 from other code.**

In [194]:
len(check)/len(df_play[df_play["user_id"] == 48798067])

0.4409448818897638